### DISCUSSION TITLE
Custom Menus

### DISCUSSION THREAD

**Discussion by RedLeafSystems** [rdbende/Sun-Valley-ttk-theme#78] [Ideas]
I would like to be able to implement custom Windows 11 styled menus in my Sun Valley application.
Something like this:
![image](https://user-images.githubusercontent.com/119130769/231814498-cbaa6a08-fa7e-42db-b912-dc76c6b0184c.png)
And like this when selected:
![image](https://user-images.githubusercontent.com/119130769/231814637-4faa501a-9720-41aa-9c41-b23f136a6c8b.png)
Is this possible with Sun Valley, and if not, would it be possible to add this as a feature?

**Comment by rdbende** [comment]
This would require creating a custom menu widget, as Tkinter's menu widget doesn't support theming.

**Reply by rdbende** [reply]
I think that's out of the scope of this project.

**Comment by RedLeafSystems** [comment]
@rdbende I disagree.
I think that Sun Valley is about styling `Tkinter` and `ttk` components, which includes menus.

**Reply by rdbende** [reply]
Ok. I thought this is a `ttk` theme, and not an extension library for Tkinter, hence its name.

That said, if anyone has a decent inplementation for a custom ttk menu eidget, I'd love to merge a PR.

**Reply by RedLeafSystems** [reply]
> Ok. I thought this is a `ttk` theme, and not an extension library for Tkinter, hence its name.
> 
> That said, if anyone has a decent inplementation for a custom ttk menu eidget, I'd love to merge a PR.

That was just what *I* was looking for! But nobody else has anything as far as I can see, and I don't know how to go about doing this.

**Comment by littlewhitecloud** [comment]
Hope the menu like this too, I think maybe can make a new repo named : "Sun-valley-ttk-widget" and make some special widgets?

**Reply by rdbende** [reply]
If someone wants it, go ahead :)) There's an ongoing effort to make the `ttkwidgets` library compatible with this theme, so would support such initiative too, but don't expect me to create these widgets.

**Comment by littlewhitecloud** [comment]
@rdbende @BenjaminTGa Here is an example. Not really good lol.
```python
from ctypes import windll
from tkinter import Tk
from tkinter.ttk import Button

from pywinstyles import change_header_color
from sv_ttk import set_theme

menu = Tk()
menu.title("menu")  # Set title
menu.iconbitmap("")  # Without this line, GetParent function could not get the hwnd
menu.minsize(175, 370)  # Set the minsize
menu.overrideredirect(1)  # Remove titlebar
menu.resizable(True, False)  # Use `resizeable(False, False)`` may make the window crash

set_theme("dark")  # Set dark theme

hwnd = windll.user32.GetParent(menu.winfo_id())  # Get hwnd

WS_THICKFRAME = 0x00040000
WS_VISIBLE = 0x10000000

windll.user32.SetWindowLongW(
    hwnd, -16, WS_THICKFRAME | WS_VISIBLE
)  # Set window to apply round frame

change_header_color(
    menu, "#1C1C1C"
)  # Set the 6px frame color as same as the window backgrorund

for text in ("New Tab", "New window", "Open", "Save", "Save as", "Save all"):  # Example
    widget = Button(
        menu,
        text=text,
    )
    widget.pack(fill="x", side="top", padx=3, pady=1)

menu.mainloop()
```

![image](https://github.com/rdbende/Sun-Valley-ttk-theme/assets/71159641/55267e1c-8cc3-4e4e-aad2-931363a2e664)

**Comment by Akascape** [comment]
@BenjaminTGa Also check this: https://github.com/Akascape/CTkMenuBar

![titlebar](https://github.com/Akascape/CTkMenuBar/assets/89206401/da6cd858-f700-476c-a2f0-93a1c6335a4d)

**Reply by rdbende** [reply]
Cool stuff, though the theme is a pretty big step-back in compared to `sv-ttk`'s style
